lib/ipmi_nm.c

```diff
@@ -0,0 +1,1834 @@
+	 * We'll use val2str from helper.c to print the error from either the
+	 * DCMI completion code dictionary or the generic IPMI codes dictionary.
+	 */
+	if (rsp->ccode) {
+		/* Try to find the code in NM dictionary */
+		descr = vs_lookup(rsp->ccode, nm_ccode_vals);
+		if (!descr) {
+			/* Fall back to generic dictionary */
+			descr = val2str(rsp->ccode, completion_code_vals);
+		}
+		lprintf(LOG_ERR, "\n    NM request failed because: %s (0x%02X)",
+		        descr, rsp->ccode);
+		goto out;
+	}
+
+	/* The code was OK, but was it an NM firmware at all? */
+	if (!nm_check_id(rsp->data)) {
+		printf("\n    No valid NM ID in response! (%06x)",
+		       ipmi24toh(rsp->data));
+		goto out;
+	}
+
+	rc = false; /* No error */
+out:
+	return rc;
+}
+
+/* Node Manager discover */
+static int
+_ipmi_nm_discover(struct ipmi_intf *intf, struct nm_discover *disc)
+{
+	struct ipmi_rq req; /* request data to send to the BMC */
+	struct ipmi_rs *rsp;
+	uint8_t msg_data[3]; /* 'raw' data to be sent to the BMC */
+
+	nm_set_id(msg_data);
+	memset(&req, 0, sizeof(req));
+	req.msg.netfn = IPMI_NETFN_OEM;
+	req.msg.cmd = IPMI_NM_GET_VERSION;
+	req.msg.data = msg_data;
+	req.msg.data_len = 3;
+	rsp = intf->sendrecv(intf, &req);
+	if (is_nm_error(rsp)) {
+		return -1;
+	}
+	memcpy(disc, rsp->data, sizeof(struct nm_discover));
+	return 0;
+}
+/* Get NM capabilities
+ *
+ * This function returns the available capabilities of the platform.
+ *
+ * returns success/failure
+ *
+ * @intf:   ipmi interface handler
+ * @caps:   fills in capability struct
+ */
+static int
+_ipmi_nm_getcapabilities(struct ipmi_intf *intf, uint8_t domain,
+                         uint8_t trigger, struct nm_capability *caps)
+{
+	struct ipmi_rq req; /* request data to send to the BMC */
+	struct ipmi_rs *rsp;
+	uint8_t msg_data[5]; /* 'raw' data to be sent to the BMC */
+
+	nm_set_id(msg_data);
+	msg_data[3] = domain;
+	msg_data[4] = trigger; /* power control policy or trigger */
+	memset(&req, 0, sizeof(req));
+	req.msg.netfn = IPMI_NETFN_OEM;
+	req.msg.cmd = IPMI_NM_GET_CAP;
+	req.msg.data = msg_data;
+	req.msg.data_len = 5;
+	rsp = intf->sendrecv(intf, &req);
+	if (is_nm_error(rsp)) {
+		return -1;
+	}
+	memcpy(caps, rsp->data, sizeof(struct nm_capability));
+	return 0;
+}
+
+static int
+_ipmi_nm_get_policy(struct ipmi_intf *intf, uint8_t domain, uint8_t policy_id,
+                    struct nm_get_policy *policy)
+{
+	struct ipmi_rq req; /* request data to send to the BMC */
+	struct ipmi_rs *rsp;
+	uint8_t msg_data[5]; /* 'raw' data to be sent to the BMC */
+
+	nm_set_id(msg_data);
+	msg_data[3] = domain;
+	msg_data[4] = policy_id;
+	memset(&req, 0, sizeof(req));
+	req.msg.netfn = IPMI_NETFN_OEM;
+	req.msg.cmd = IPMI_NM_GET_POLICY;
+	req.msg.data = msg_data;
+	req.msg.data_len = 5;
+	rsp = intf->sendrecv(intf, &req);
+	if (is_nm_error(rsp)) {
+		return -1;
+	}
+	memcpy(policy, rsp->data, sizeof(struct nm_get_policy));
+	return 0;
+}
+static int
```

Comment:
I'm assuming since this is just moved code, there'll be possibly follow-on for style conformance?